File: Backend/QueryBuilderAPI/Models/DatabaseModels.cs
namespace QueryBuilderAPI.Models
{
    public class SchemaInfo
    {
        public string SchemaName { get; set; } = string.Empty;
    }

    public class TableInfo
    {
        public string SchemaName { get; set; } = string.Empty;
        public string TableName { get; set; } = string.Empty;
    }

    public class ColumnInfo
    {
        public string SchemaName { get; set; } = string.Empty;
        public string TableName { get; set; } = string.Empty;
        public string ColumnName { get; set; } = string.Empty;
        public string DataType { get; set; } = string.Empty;
        public bool IsNullable { get; set; }
        public bool IsPrimaryKey { get; set; }
        public bool IsForeignKey { get; set; }
    }

    public class ForeignKeyInfo
    {
        public string ConstraintName { get; set; } = string.Empty;
        public string SourceSchema { get; set; } = string.Empty;
        public string SourceTable { get; set; } = string.Empty;
        public string SourceColumn { get; set; } = string.Empty;
        public string TargetSchema { get; set; } = string.Empty;
        public string TargetTable { get; set; } = string.Empty;
        public string TargetColumn { get; set; } = string.Empty;
    }

    public class QueryRequest
    {
        public List<SelectedColumn> Columns { get; set; } = new();
        public List<JoinInfo> Joins { get; set; } = new();
        public List<WhereCondition> WhereConditions { get; set; } = new();
        public string? OrderBy { get; set; }
        public string? OrderDirection { get; set; }
        public int? Limit { get; set; }
    }

    public class SelectedColumn
    {
        public string SchemaName { get; set; } = string.Empty;
        public string TableName { get; set; } = string.Empty;
        public string ColumnName { get; set; } = string.Empty;
        public string? Alias { get; set; }
    }

    public class JoinInfo
    {
        public string JoinType { get; set; } = "INNER"; // INNER, LEFT, RIGHT, FULL
        public string LeftSchema { get; set; } = string.Empty;
        public string LeftTable { get; set; } = string.Empty;
        public string LeftColumn { get; set; } = string.Empty;
        public string RightSchema { get; set; } = string.Empty;
        public string RightTable { get; set; } = string.Empty;
        public string RightColumn { get; set; } = string.Empty;
    }

    public class WhereCondition
    {
        public string SchemaName { get; set; } = string.Empty;
        public string TableName { get; set; } = string.Empty;
        public string ColumnName { get; set; } = string.Empty;
        public string Operator { get; set; } = "="; // =, !=, >, <, >=, <=, LIKE, IN
        public string Value { get; set; } = string.Empty;
        public string? LogicalOperator { get; set; } // AND, OR
    }

    public class QueryResult
    {
        public string GeneratedSql { get; set; } = string.Empty;
        public List<Dictionary<string, object?>> Data { get; set; } = new();
        public int TotalRows { get; set; }
    }
}

File: Frontend/src/app/models/database.models.ts
export interface SchemaInfo {
  schemaName: string;
}

export interface TableInfo {
  schemaName: string;
  tableName: string;
}

export interface ColumnInfo {
  schemaName: string;
  tableName: string;
  columnName: string;
  dataType: string;
  isNullable: boolean;
  isPrimaryKey: boolean;
  isForeignKey: boolean;
}

export interface ForeignKeyInfo {
  constraintName: string;
  sourceSchema: string;
  sourceTable: string;
  sourceColumn: string;
  targetSchema: string;
  targetTable: string;
  targetColumn: string;
}

export interface SelectedColumn {
  schemaName: string;
  tableName: string;
  columnName: string;
  alias?: string;
}

export interface JoinInfo {
  joinType: string; // 'INNER' | 'LEFT' | 'RIGHT' | 'FULL'
  leftSchema: string;
  leftTable: string;
  leftColumn: string;
  rightSchema: string;
  rightTable: string;
  rightColumn: string;
}

export interface WhereCondition {
  schemaName: string;
  tableName: string;
  columnName: string;
  operator: string; // '=' | '!=' | '>' | '<' | '>=' | '<=' | 'LIKE' | 'IN'
  value: string;
  logicalOperator?: string; // 'AND' | 'OR'
}

export interface OrderByClause {
  schemaName: string;
  tableName: string;
  columnName: string;
  direction: string; // 'ASC' | 'DESC'
}

export interface GroupByClause {
  schemaName: string;
  tableName: string;
  columnName: string;
}

export interface QueryRequest {
  columns: SelectedColumn[];
  joins: JoinInfo[];
  whereConditions: WhereCondition[];
  orderBy?: OrderByClause[];
  groupBy?: GroupByClause[];
  havingConditions?: WhereCondition[];
  limit?: number;
  offset?: number;
}

export interface QueryResult {
  generatedSql: string;
  data: any[];
  totalRows: number;
}
